pkg/interface/publish/src/js/components/lib/subscriber-item.js

```diff
@@ -1,12 +1,149 @@
-import React, { Component } from 'react'
-
-//TODO fill sigil/avatar + name from props
+import React, { Component } from 'react';
+import { Link } from 'react-router-dom';
 
 export class SubscriberItem extends Component {
+  constructor(props) {
+    super(props);
+
+    this.toggleOptions = this.toggleOptions.bind(this);
+    this.handleClickOutside = this.handleClickOutside.bind(this);
+    this.ban = this.ban.bind(this);
+    this.unban = this.unban.bind(this);
+    this.removeWriter = this.removeWriter.bind(this);
+    this.addWriter = this.addWriter.bind(this);
+    this.state = {
+      optionsSelected: false
+    }
+  }
+
+  componentDidMount() {
+    document.addEventListener('mousedown', this.handleClickOutside);
+  }
+
+  componentWillUnmount() {
+    document.removeEventListener('mousedown', this.handleClickOutside);
+  }
+
+  handleClickOutside(evt) {
+    if (this.optsList && !this.optsList.contains(evt.target) &&
+        this.optsButton && !this.optsButton.contains(evt.target)) {
+      this.setState({optionsSelected: false});
+    }
+  }
+
+  ban() {
+    let action = {
+      add: {
+        members: [this.props.who],
+        path: this.props.readPath,
+      }
+    }
+    this.setState({optionsSelected: false}, () => {
+      window.api.action("group-store", "group-action", action);
+    });
+  }
+
+  unban() {
+    let action = {
+      remove: {
+        members: [this.props.who],
+        path: this.props.readPath,
+      }
+    }
+    this.setState({optionsSelected: false}, () => {
+      window.api.action("group-store", "group-action", action);
+    });
+  }
+
+  removeWriter() {
+    let action = {
+      remove: {
+        members: [this.props.who],
+        path: this.props.writePath,
+      }
+    }
+    this.setState({optionsSelected: false}, () => {
+      window.api.action("group-store", "group-action", action);
+    });
+  }
+
+  addWriter() {
+    let action = {
+      add: {
+        members: [this.props.who],
+        path: this.props.writePath,
+      }
+    }
+    this.setState({optionsSelected: false}, () => {
+      window.api.action("group-store", "group-action", action);
+    });
+  }
+
+  toggleOptions() {
+    this.setState({optionsSelected: !this.state.optionsSelected});
+  }
+
   render() {
+    let display = (this.state.optionsSelected)
+      ? "block" : "none";
+
+    let width = 0;
+    let options = [];
+    if (this.props.section === 'participants') {
+      if (this.props.readPath === this.props.writePath) {
+        width = 258;
+        let url = `/~contacts${this.props.writePath}`;
+        options = [
+          <a key={0} className="tl pointer" href={url}>
+            Manage this group in the contacts view
```

Comment:
Flagging this as something where the API logic and view logic should be broken apart. Pass in function calls to APIs as props, and options as props, rather than hardcoding